docs/dsl/example.md

```diff
@@ -0,0 +1,438 @@
+
+**patients**
+
+| patient | age | sex |
+| - | - | - |
+| 1234 | 42 | f |
+| 2345 | 51 | m |
+| 3456 | 83 | f |
+| 4567 | 61 | m |
+| 5678 | 29 | f |
+
+**registrations**
+
+| id | patient | practice | start_date | end_date |
+| - | - | - | - | - |
+| 78 | 1234 | A | 1989-03-12 | 2021-04-13 |
+| 79 | 1234 | B | 2012-04-13 | - |
+| 80 | 2345 | C | 1999-12-01 | - |
+| 81 | 3456 | D | 2001-12-03 | 2017-09-09 |
+| 82 | 4567 | E | 2019-06-01 | 2015-09-01 |
+| 83 | 4567 | F | 2015-10-25 | - |
+| 84 | 5678 | G | 2019-02-12 | - |
+
+
+**immunisations**
+
+| id | patient | code | date |
+| - | - | - | - |
+| 43 | 1234 | pf | 2021-03-19 |
+| 44 | 4567 | pox | 2021-10-10 |
+| 45 | 4567 | az | 2021-06-23 |
+| 46 | 1234 | flu | 2021-10-15 |
+| 47 | 1234 | pf | 2021-07-04 |
+| 48 | 2345 | flu | 2021-03-25 |
+| 49 | 2345 | flu | 2021-08-19 |
+| 50 | 5678 | pox | 2021-10-11 |
+| 51 | 4567 | flu | 2021-12-14 |
+| 52 | 3456 | pf | 2021-09-09 |
+| 53 | 2345 | flu | 2021-11-02 |
+| 54 | 5678 | pf | 2021-07-01 |
+| 55 | 3456 | pox | 2021-11-03 |
+| 56 | 3456 | az | 2021-04-01 |
+| 57 | 5678 | az | 2021-12-30 |
+| 58 | 5678 | pf | 2021-07-02 |
+
+**patients** is a `PatientFrame`, while **registrations** and **immunisations** are `EventFrame`s.
+
+---
+
+A `PatientFrame` represents a collection of rows with at most one row per patient.  An `EventFrame` represents as collection of rows with multiple rows per patient.
+
+An `EventFrame`/`PatientFrame` can be filtered to produce another `EventFrame`/`PatientFrame`, and you should think of them as being like unevaluated Django Querysets, in that they can be combined and acted on to produce something that can be turned into SQL.
+
+Ultimately, we transform an `EventFrame`/`PatientFrame` into a `PatientSeries`, through filtering, sorting, aggregating, and selecting a column.
+
+A `PatientSeries` represents a collection of values with one value per patient.
+
+A `PatientSeries` can be assigned to a cohort, in which case a column with the `PatientSeries`'s values will be appended to the cohort's input.csv.
+
+---
+
+This diagram attempts to explain how a `EventFrame`/`PatientFrame` is transformed into a
+`PatientSeries`.
+
+The nodes are Python classes and the arrows are methods on those classes.
+
+```
+             +---+
+             |   V
+      filter |  PatientFrame
+             |   |   |
+             +---+   | select_column
+                     V
+                PatientSeries
+
+
+             +---+
+             |   V
+      filter |  EventFrame -----------------------+
+             |   |   |                            |
+             +---+   | sort_by                    |
+                     V                            |
+             SortedEventFrame                     |
+                     |                            |
+        +---+        | (first/last)_for_patient   | (count/exists)_for_patient
+        |   V        V                            |
+ filter |  AggregatedEventFrame                   |
```

Comment:
chore: The diagram here and associated class names in a tutorial would need updating.